Core interfaces for system tags and its manager

PVince81 opened the pull request:
@nickvergessen as discussed

I'm not sure if accepting an array of tagIds everywhere is a good idea, but it's certainly convenient for the API consumer.

I purposefully made `ISystemTag` interface read-only. The only value that can be edited is anyway the tag name. Changing visibility is not possible because that could overlap with another tag...

CC @blizzz @icewind1991 @DeepDiver1975

Commit:
Core interfaces for system tags and its manager

PVince81 commented:
For https://github.com/owncloud/core/issues/16325

PVince81 commented on lib/public/isystemtagsmanager.php:
```diff
@@ -0,0 +1,154 @@
+	 * @return bool Returns false on error
+	 * @since 9.0.0
+	 */
+	public function delete($tagIds);
```
I forgot to fix the comment of this one, will do at the next push opportunity.

blizzz commented on lib/public/isystemtag.php:
```diff
@@ -0,0 +1,60 @@
+	/**
+	 * Returns the tag id
+	 *
+	 * @return string id
```
@since is also required for all methods

PVince81 replied:
bääh... I just removed them.

I'll add them back :smile: 
Thanks

nickvergessen replied:
id is auto increment and therefor integer?

PVince81 replied:
I thought string because we wanted to be able to support string ids in general for alternative implementations (like for file cache). The default implementation can still use integers and autoincrement.

nickvergessen commented on lib/public/isystemtagsmanager.php:
```diff
@@ -0,0 +1,152 @@
+ *
+ */
+
+namespace OCP;
```
OCP\SystemTag\IManager ?
Stop polluting the root :stuck_out_tongue_winking_eye:

nickvergessen commented on lib/public/isystemtagsmanager.php:
```diff
@@ -0,0 +1,152 @@
+	 * @return bool true if the condition set by $all is matched, false
+	 * otherwise
+	 */
+	public function hasTags($objId, $objectType, $tagIds, $all = true);
```
I need this method the otherway around `$objIds` and `$tagId`, so multiple objects and one tag

nickvergessen replied:
but maybe it's good to have both

PVince81 replied:
ah yes, my mistake

nickvergessen commented:
> The only value that can be edited is anyway the tag name.

Wrong, that can also create a conflict in theory :see_no_evil: 
But you will get a constraint violation when trying it, if the new name is already in use,
so just catch that exception.

nickvergessen commented on lib/public/isystemtag.php:
```diff
@@ -0,0 +1,60 @@
+ *
+ */
+
+namespace OCP;
```
`OCP\SystemTag\ITag`

nickvergessen commented on lib/public/isystemtagsmanager.php:
```diff
@@ -0,0 +1,152 @@
+	public function unassignTags($objId, $objectType, $tagIds);
+
+	/**
+	 * Replaces all tags assigned to the given object with the new given tag ids.
```
this sounds dangerous, what's the usecase?
I'd prefer to do `unassignAllTags($objId, $objectType)` + `assignTags($objId, $objectType, $tagIds)` instead

PVince81 replied:
The use case is simply: the web UI contains a list of tags that are assigned to the file. When we update the list, we need to pass the whole list back to the backend, and set it there.

If we don't provide this method, then we need to do diffing to find out which tags have been assigned/unassigned from the original list, like we did here: https://github.com/owncloud/core/blob/master/apps/dav/lib/connector/sabre/tagsplugin.php#L181

Also note that PROPPATCH with Webdav would also need the an absolute list of tags through a single array property "oc:systemtags". There is no way to insert an element in a subproperty.

nickvergessen replied:
Then I see concurrency problems:
1. Open sidebar for user1
2. Open sidebar for user2
3. Set Tag1 as user1 => Sends [Tag1]
4. Set Tag2 as user2 => Sends [Tag2] which removes [Tag1], although unintended?

So I'd say diff it and then call unassign and assign split in two calls

PVince81 replied:
The concurrency might still happen because to be able to do the diff you need to "get all tags assigned to this file" and then do the diff and then assign/unassign. If between that a concurrent process does the same, they'll have a different list.

Best here is to use the transactional locking on that specific file before doing tags operations. The locking would be triggered by whichever Controller class is handling the request.

PVince81 replied:
I agree to remove this method.

nickvergessen replied:
> Best here is to use the transactional locking on that specific file before doing tags operations. The locking would be triggered by whichever Controller class is handling the request.

No, you are misunderstanding. But yes, just remove it.

nickvergessen commented on lib/public/isystemtagsmanager.php:
```diff
@@ -0,0 +1,152 @@
+	 *
+	 * @return array array of \OCP\ISystemTag
+	 */
+	public function getAllTags($visibleOnly = false);
```
maybe it would also be nice to have filter options for the name in this, e.g. for type ahead like in the search dropdown, instead of preloading all of them?

DeepDiver1975 commented on lib/public/isystemtagsmanager.php:
```diff
@@ -0,0 +1,152 @@
+	 *
+	 * @param array|string $tagIds The ID or array of IDs of the tags to retrieve
+	 *
+	 * @return array array of \OCP\ISystemTag
```
\OCP\ISystemTag[]

DeepDiver1975 commented on lib/public/isystemtagsmanager.php:
```diff
@@ -0,0 +1,152 @@
+	 *
+	 * @return \OCP\ISystemTag
+	 */
+	public function getOrCreateTag($tagName, $userVisible, $userAssignable);
```
I would not mix the operations - there is a explicit create and an explicit get which returns an error

DeepDiver1975 commented on lib/public/isystemtagsmanager.php:
```diff
@@ -0,0 +1,152 @@
+	 * @param string $tagId tag id
+	 * @param string $newName the new tag name
+	 */
+	public function renameTag($tagId, $newName);
```
what about having a more general update method?

PVince81 replied:
Maybe. Keep it mind that the name is the only thing that we can change.
Visibility flags cannot be changed because it might overwrite other existing tags. (a tag name can exist twice with different visibilities...)

nickvergessen replied:
can happen for names as well, as I mentioned else where

1. Create Tag `tag1` vis + asg
2. Create Tag `tag2` vis + asg
3. Rename Tag `tag1` vis + asg to `tag2` vis + asg

So no, we can not change names really either or we allow changing anything

PVince81 replied:
Right, because the unique key is "tag name" + "visibility" + "assignable" ?

I think it was an explicit request to be able to rename/edit tags. If that's not possible, then we'd need a bulk operation that creates the new tag, then assigns the new one and deletes the old one. That operation could still be done within the update method.

What do you think ?

nickvergessen replied:
just add a method that tries changing it? If the other key already exists throw an exception, if it does not succeed?

I don't see the problem really

PVince81 replied:
yes, I'll rename this to `update($tagId, $tagName)` but without the visibility flags.
Or do we want to make these updatable as well, using the bulk operation trick proposed above ?

nickvergessen replied:
No need for bulk operation?

It's a single query:
```sql
UPDATE oc_systemtag SET name = ?, visibility = ?, assignability = ? WHERE tag_id = ?
```

When the query fails with constraint violation, because the name+vis+asg combination already exists, throw Exception, if it passes you changed the tag correctly?!

PVince81 replied:
Right, that makes sense

DeepDiver1975 commented on lib/public/isystemtagsmanager.php:
```diff
@@ -0,0 +1,152 @@
+ *
+ * @since 9.0.0
+ */
+interface ISystemTagManager {
```
should we split the managers? one for assignment handling and the other for explicit tag creation/modification/deletion?

nickvergessen replied:
I think it makes somewhat sense to split those. helps keeping files smaller.

PVince81 commented:
All comments adressed, except the "string id" one.

Now another topic: exceptions.
See https://github.com/owncloud/core/pull/20563/files

Currently we don't have a global generic "NotFoundException" and "AlreadyExistsException".

Should we add `\OCP\NotFoundException` and make `\OCP\Files\NotFoundException` inherit it ? (same for AlreadyExistsException)

PVince81 commented:
Also I wonder if we should keep the plural `tagIds` in many cases, because that would require a `select ... where tagid in (...)` and requires the magic chunking logic for sqlite and co...

nickvergessen commented on lib/public/systemtag/isystemtagobjectmapper.php:
```diff
@@ -0,0 +1,106 @@
+	/**
+	 * Get a list of objects tagged with $tagIds.
+	 *
+	 * Throws an exception if the tag could not be found.
```
should be `@throws`

nickvergessen commented on lib/public/systemtag/isystemtagobjectmapper.php:
```diff
@@ -0,0 +1,106 @@
+	 *
+	 * @param string $objId object id
+	 * @param string $objectType object type
+	 * @param string|array $tagIds tag id or array of tag ids to unassign
```
to assign

nickvergessen commented on lib/public/systemtag/isystemtagobjectmapper.php:
```diff
@@ -0,0 +1,106 @@
+	/**
+	 * Checks whether the given objects have the given tag.
+	 *
+	 * @param string $objIds object ids
```
array?

nickvergessen commented on lib/public/systemtag/isystemtagobjectmapper.php:
```diff
@@ -0,0 +1,106 @@
+	 *
+	 * @param string $objIds object ids
+	 * @param string $objectType object type
+	 * @param string $tagIds tag id or array of tag ids to check
```
Description of the method says "have the given tag." so maybe only allow one tag

nickvergessen commented on lib/public/systemtag/isystemtagobjectmapper.php:
```diff
@@ -0,0 +1,106 @@
+	public function getObjectIdsForTags($tagIds, $objectType);
+
+	/**
+	 * Assign the given tags to the given object.
```
assign and unassign should `@throw` as well when the tag ID is invalid

PVince81 replied:
this implies that we add a validity check ? (well, if we had foreign keys it would be implicit)

nickvergessen replied:
But we don't and so I'd say yes, call getTagById and see if that throws

PVince81 replied:
:white_check_mark: resolved

nickvergessen commented on lib/public/systemtag/isystemtagsmanager.php:
```diff
@@ -0,0 +1,109 @@
+	 *
+	 * @return \OCP\SystemTag\ISystemTag system tag
+	 *
+	 * @throws \OCP\Files\NotFoundException if tag does not exist
```
Wrong exception, this is not a File, so don't try to use that exception

PVince81 replied:
I know, hence my question in the ticket about creating new exceptions directly in `\OCP`

nickvergessen commented on lib/public/systemtag/isystemtagsmanager.php:
```diff
@@ -0,0 +1,109 @@
+	 *
+	 * @return \OCP\SystemTag\ISystemTag system tag
+	 *
+	 * @throws \OCP\Files\AlreadyExistsException if tag already exists
```
Same here

nickvergessen commented on lib/public/systemtag/isystemtagsmanager.php:
```diff
@@ -0,0 +1,109 @@
+	 * @param bool $userVisible whether the tag is visible by users
+	 * @param bool $userAssignable whether the tag is assignable by users
+	 *
+	 * @throws \OCP\Files\AlreadyExistsException if tag already exists
```
same here

nickvergessen replied:
Should also throw if tag_id was not found

nickvergessen commented on lib/public/systemtag/isystemtagsmanager.php:
```diff
@@ -0,0 +1,109 @@
+	 *
+	 * @param string|array $tagIds array of tag ids
+	 *
+	 * @throws \OCP\Files\NotFoundException if tag does not exist
```
same here

nickvergessen replied:
arguable if this should throw at all, or just accept it silently that it's job has already been completed

PVince81 commented:
- [ ] TODO: create specific exceptions (namespaced) for system tags ("not found" and "already exists")

PVince81 commented:
I added new exception `TagNotFoundException` and `TagAlreadyExistsException` and fixed all the comments.

Please have another look and thanks for the reviews so far :smile:

nickvergessen commented on lib/public/systemtag/isystemtagobjectmapper.php:
```diff
@@ -0,0 +1,116 @@
+	 * @return bool true if the condition set by $all is matched, false
+	 * otherwise
+	 *
+	 * @throws \OCP\SystemTag\TagNotFoundException if at least one of the
```
its only one, so "at least one" is not the right word, but can fix later

nickvergessen commented on lib/public/systemtag/isystemtagsmanager.php:
```diff
@@ -0,0 +1,111 @@
+	 *
+	 * @param string|array $tagIds array of tag ids
+	 *
+	 * @throws \OCP\SystemTag\TagNotFoundException if tag does not exist
```
*did not exist

nickvergessen commented:
:+1:

PVince81 commented:
Can we merge this interface @DeepDiver1975 ?
